Data dependency between layers + snapping fix

mhugo opened the pull request:
Following #3306, this patch adds a way for the user to declare data dependencies between layers, where data of a layer may be modified when other layers are modified.

Dependencies are declared thanks to new methods on QgsProject. Layers are registered for modification signals (featureAdded / featureDelete / geometryChanged) and will trigger dataChanged()

Other parts of the code that should know which caches to update can connect to the dataChanged signal. This is done for the snapping on QgsPointLocator

UI side, there is a new tab in the vector layer property that allows to select which layers depends on which others.

wonder-sk commented on src/core/qgspointlocator.cpp:
```diff
@@ -628,6 +628,7 @@ QgsPointLocator::QgsPointLocator( QgsVectorLayer* layer, const QgsCoordinateRefe
   connect( mLayer, SIGNAL( featureAdded( QgsFeatureId ) ), this, SLOT( onFeatureAdded( QgsFeatureId ) ) );
   connect( mLayer, SIGNAL( featureDeleted( QgsFeatureId ) ), this, SLOT( onFeatureDeleted( QgsFeatureId ) ) );
   connect( mLayer, SIGNAL( geometryChanged( QgsFeatureId, QgsGeometry& ) ), this, SLOT( onGeometryChanged( QgsFeatureId, QgsGeometry& ) ) );
+  connect( mLayer, SIGNAL( dataChanged() ), this, SLOT( rebuildIndex() ) );
```
It should be enough to destroy the index rather than rebuilding it immediately - it will be built when necessary later.

mhugo replied:
Exact. Thanks !

wonder-sk commented:
Hi Hugo

Looking at the code, I am wondering if the storage of dependent layers would not be better done within QgsMapLayer instead of QgsProject:
- it is a per layer property
- within QgsMapLayer one would not need to have emitDataChanged() method (it hurts my heart to see emitXyz() methods as it breaks the expectation that signals are emitted by the object)
- within QgsMapLayer one could better define the behavior specific for layer type (which signals to handle) and it would allow plugin layers to be used too

Other notes:
- How do we ensure that no circular dependencies are made? (that could lead to infinite loop of updates?) For vector joins we already have a simple circular dependency check.
- What if a layer that another layer depends on is missing? (e.g. on project load or if a layer is removed). I understand that your concern here is mainly to "just" invalidate cached data, but for example virtual layers require other layers to be present in order to work - should we distinguish between "hard" and "soft" dependencies somehow?

mhugo commented:
Hi Martin, thanks for your feedback.

> I am wondering if the storage of dependent layers would not be better done within QgsMapLayer instead of QgsProject

I agree for the storage in QgsMapLayer rather than in QgsProject. I thought it would lead to too much signal connections for each layer (waiting for dependent layers to be loaded), but I realize it can be done during project loading ... I'll change that.

> How do we ensure that no circular dependencies are made? (that could lead to infinite loop of updates?)

Unless dataChanged() triggers a featureAdded() / featureDeleted() / geometryChanged(), which is not the case, there cannot be infinite loop, I think.

> What if a layer that another layer depends on is missing?

No connection will be made or the connection will be lost.
But ... there could be the following problem: what to do if a layer A depends on B and is loaded without B present (in embedded layer for example) and then B is loaded later ? That means there should be something waiting for dependent layers not yet resolved to arrive ... perhaps in QgsProject when loading embedded layers.

> should we distinguish between "hard" and "soft" dependencies somehow?

There is already a layerDependencies() on QgsVectorLayer, which is read-only, these "hard" dependent layers are given by the provider (only virtual provider so far). Are there cases where the user wants to declare "hard" dependencies ?
I was thinking about adding a layerDataDependencies() (or layerChangeDependencies ? ) for "soft" dependencies.

wonder-sk commented:
> > How do we ensure that no circular dependencies are made? (that could lead to infinite loop of  updates?)
>
> Unless dataChanged() triggers a featureAdded() / featureDeleted() / geometryChanged(), which is not the case, there cannot be infinite loop, I think.

I see... but if layer B depends on layer A, shouldn't also A's dataChanged() signal trigger B's dataChanged() signal? The dataChanged() signal may be also be coming from A's data provider (when shapefile is repacked, when subset string changes, when source has been updated (e.g. GRASS)). And if dataChanged() signals are connected between layers, then we need cycle detection too...

> But ... there could be the following problem: what to do if a layer A depends on B and is loaded without B present (in embedded layer for example) and then B is loaded later ? That means there should be something waiting for dependent layers not yet resolved to arrive ... perhaps in QgsProject when loading embedded layers.

There are already some bits of code in QgsProject that take care of similar things, e.g. to establish joins between layers after loading them...

> There is already a layerDependencies() on QgsVectorLayer, which is read-only, these "hard" dependent layers are given by the provider (only virtual provider so far).

Ah right - I must have missed that one.

> Are there cases where the user wants to declare "hard" dependencies ?

Not sure really.

> I was thinking about adding a layerDataDependencies() (or layerChangeDependencies ? ) for "soft" dependencies.

My thought about hard/soft dependencies was mainly about whether we should not have a common approach for both of them, e.g. dependent layers could be a list of layers, each layer having flags about the dependency details (e.g. "hard" vs "soft"). Some of the dependencies could be implied (read-only, given by provider), others would be configured by users...

wonder-sk commented on src/app/qgsvectorlayerproperties.cpp:
```diff
@@ -681,6 +698,16 @@ void QgsVectorLayerProperties::apply()
+      deps << layer->layerId();
+  }
+  QgsProject::instance()->setLayerDependencySources( mLayer->id(), deps );
+  mLayer->emitDataChanged();
```
Is it necessary to emit data changed signal here? I would have guessed that just configuring a dependency does not need to emit this signal - confirming the dialog will always destroy the caches even if not necessary...

Commit:
Snapping: destroy index on dataChanged signal

Snapping caches on layers that have been changed by the provider or
by external sources are now invalidated.

Commit:
[Feature] Data dependencies between layers

This allows to declare data dependencies between layers. A data
dependency occurs when a data modification in a layer, not by direct
user manipulation may modify data of other layers.
This is the case for instance when geometry of a layer is updated by a
database trigger after modification of another layer's geometry.

mhugo commented:
> I see... but if layer B depends on layer A, shouldn't also A's dataChanged() signal trigger B's dataChanged() signal? The dataChanged() signal may be also be coming from A's data provider (when shapefile is repacked, when subset string changes, when source has been updated (e.g. GRASS)). And if dataChanged() signals are connected between layers, then we need cycle detection too...

You're right. If updateChanged() are chained, then a cycle detection is needed. And indeed, it makes sense to chain them. I've added a cycle detection, thanks.

> My thought about hard/soft dependencies was mainly about whether we should not have a common approach for both of them, e.g. dependent layers could be a list of layers, each layer having flags about the dependency details (e.g. "hard" vs "soft"). Some of the dependencies could be implied (read-only, given by provider), others would be configured by users...

I don't have a strong preference here. But I don't see much commonalities between these two kinds of dependencies. Data dependencies introduced here actually share more code with vector joins than with loading dependencies.

wonder-sk commented on src/core/qgsvectorlayer.cpp:
```diff
@@ -4101,3 +4121,49 @@ QSet<QString> QgsVectorLayer::layerDependencies() const
+
+QSet<QString> QgsVectorLayer::dataDependencies() const
+{
+  return layerDependencies() + mDataDependencies;
```
Aren't layers from layerDependencies() already included from previous setDataDependencies() call?

wonder-sk commented on src/core/qgsproject.cpp:
```diff
@@ -975,6 +983,17 @@ void QgsProject::onMapLayersAdded( const QList<QgsMapLayer*>& layers )
+      if ( deps.contains( layer->id() ) )
+      {
+        // reconnect to change signals
+        it.value()->setDataDependencies( deps );
```
I am not sure if I understand why this needs to happen. Aren't dependencies dismissed in existing layers if they were not found previously?

mhugo replied:
No the dependency is still there, but no connection is made if the other layer is not there yet.

This is to cover the case where you have a layer A that depends on B and you load them one after the other (embbeded layers for instance)
* A is loaded, B is not there -> no signal connection
* B is loaded afterward and it will reset A's connections, B is there now and the connection can be made

And ... I even realize this has to be done recursively (if A -> B -> C and they are loaded in that order)

wonder-sk commented:
Hi Hugo - thanks for the updates!

> I don't have a strong preference here. But I don't see much commonalities between these two kinds of dependencies. Data dependencies introduced here actually share more code with vector joins than with loading dependencies.

Looking into the code, there is a fair amount of dependencies in code between "layer" and "data" dependencies. My main concern is that in few days I will probably not remember the difference between "layer" and "data" dependencies and what is their relationship. If we had just one kind of dependencies, with a flag for each layer ("must be present", "modifies data"), it would be probably easier to figure what is going on?

mhugo commented:
> My main concern is that in few days I will probably not remember the difference between "layer" and "data" dependencies and what is their relationship. If we had just one kind of dependencies, with a flag for each layer ("must be present", "modifies data"), it would be probably easier to figure what is going on? 

Do you mean something like that ? 
```c++
class QgsMapLayerDependency
{
  enum Type
  {
    PresenceDependency = 1,
    DataDependency     = 2
  };

  enum Origin
  {
    FromProvider = 0,
    FromUser     = 1
  };
    
  Type type() const;
  Origin origin() const;
  QString layerId() const;
};

class QgsMapLayerDependencies : public QSet<QgsMapLayerDependency> //<- just for illustration
{
public:
  QgsMapLayerDependencies selectByType( QgsMapLayerDependency::Type type ) const;
  QgsMapLayerDependencies selectByOrigin( QgsMapLayerDependency::Origin origin ) const;
}
```

Commit:
Rework layer dependencies

A new class QgsMapLayerDependency allows to represent different kinds
of dependencies between layers.

mhugo commented:
Back on this topic. I've added a unit test and reworked a little bit dependencies. There is a new class QgsMapLayerDependency to handle the two kinds of dependencies. @wonder-sk
